middleware/domain/include/domain/discovery/state.h

```diff
@@ -23,17 +23,114 @@ namespace casual
    {
       namespace state
       {
-         enum struct Runlevel : short
+         namespace metric
          {
-            running,
-            shutdown,
-         };
-         std::string_view description( Runlevel value);
+            namespace message
+            {
+               namespace detail
+               {
+                  namespace count
+                  {
+                     enum class Tag : short
+                     {
+                        received,
+                        sent
+                     };
+
+                     template< Tag tag, common::message::Type type>
+                     struct basic_count
+                     {
+                        static void increment() noexcept { ++m_count;}
+                        static auto value() noexcept { return m_count;}
+                     private:
+                        static platform::size::type m_count;
+                     };
+
+                     template< Tag tag, common::message::Type type>
+                     platform::size::type basic_count< tag, type>::m_count = {};
+
+
+                     template< common::message::Type type>
+                     using Receive = basic_count< count::Tag::received, type>;
+
+                     template< common::message::Type type>
+                     using Send = basic_count< count::Tag::sent, type>;
+
+                  } // count
+
+                  template< count::Tag tag, typename C, typename Message>
+                  constexpr auto create( C creator, const Message&)
+                  {
+                     constexpr auto type = common::message::type< common::traits::remove_cvref_t< Message>>();
+                     return creator( 
+                        type, 
+                        detail::count::basic_count< tag, type>::value());
+                  }
+
+                  template< count::Tag tag, typename C, typename... Messages>
+                  constexpr auto counts( C creator, Messages... messages) noexcept
+                  {
+                     using value_type = decltype( creator( common::message::Type{}, platform::size::type{}));
+
+                     std::vector< value_type> result;
+                     result.reserve( sizeof...( Messages));
+
+                     ( ... , result.push_back( detail::create< tag>( creator, messages)) );
```

Comment:
How does this work? I assume it unpacks into a push_back for each Messages parameter but the syntax is completely new to me.

Reply:
A feature in c++17, fold-expressions https://en.cppreference.com/w/cpp/language/fold. Not often that you need it, but when you do, it's awesome!